### ISSUE TITLE
Minitest throws error when using refinements to patch an ActiveSupport extension method

### ISSUE THREAD

**Issue by erose** [seattlerb/minitest#671]
When running `rake test` since upgrading to Rails 5, I'm getting the following error (originating from `require minitest/mock` line):

```
NameError: undefined method `to_query' for class `Minitest::Mock'
Did you mean?  to_query
```

I'm able to reproduce with a single file, outside of the Rails context:

```
require 'active_support'
require 'active_support/core_ext'

module ObjectExtensions
  refine ::Object do
    def to_query
      'bar'
    end
  end
end

require 'minitest/mock'
```

Running `ruby file.rb` results in the error. This file is AFAIK fully minimal. Note that we're not even using the extension.

Thanks for all your work on Minitest! I really appreciate it.

**Comment by zenspider** [comment]
I cannot reproduce the error with the code given. Please provide more details / steps including the actual commands run and error output.

**Comment by blowmage** [comment]
@erose Is it possible some other library is in play? The `GHFMarkdown` has to come from somewhere.

**Comment by zenspider** [comment]
The what? I'm confused...

**Comment by blowmage** [comment]
The error is a missing method with a pretty crazy name. I bet whatever library that creates that method is needed in order to reproduce the error. And also the cause of the error. Some markdown library that adds GitHub-style fenced code markup, I imagine.

**Comment by erose** [comment]
@blowmage: Hmm, that's not the error that I get at all. I get exactly the error I described:

```
NameError: undefined method `to_query' for class `Minitest::Mock'
Did you mean?  to_query
```

@zenspider: My steps to reproduce are:

  * Create a new file `test.rb` whose contents are as above.
  * Run `ruby test.rb`.

Maybe we have different versions of certain dependencies.

  * Output of `ruby -v`
    * `ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]`
  * Output of `gem list activesupport`
    * `activesupport (5.0.0, 4.2.7.1)`
  * Output of `gem list minitest`
    * `minitest (5.10.1, 5.9.0, 5.8.3)`

And here is the full stack trace, in case that's helpful:

```
/home/eli/.rvm/gems/ruby-2.3.1@shearwater/gems/minitest-5.10.1/lib/minitest/mock.rb:27:in `undef_method': undefined method `to_query' for class `Minitest::Mock' (NameError)
Did you mean?  to_query
	from /home/eli/.rvm/gems/ruby-2.3.1@shearwater/gems/minitest-5.10.1/lib/minitest/mock.rb:27:in `block in <class:Mock>'
	from /home/eli/.rvm/gems/ruby-2.3.1@shearwater/gems/minitest-5.10.1/lib/minitest/mock.rb:26:in `each'
	from /home/eli/.rvm/gems/ruby-2.3.1@shearwater/gems/minitest-5.10.1/lib/minitest/mock.rb:26:in `<class:Mock>'
	from /home/eli/.rvm/gems/ruby-2.3.1@shearwater/gems/minitest-5.10.1/lib/minitest/mock.rb:10:in `<module:Minitest>'
	from /home/eli/.rvm/gems/ruby-2.3.1@shearwater/gems/minitest-5.10.1/lib/minitest/mock.rb:3:in `<top (required)>'
	from /home/eli/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from /home/eli/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from /home/eli/minimal-examples/minitest.rb:16:in `<main>'
```

**Comment by utilum** [comment]
@erose I can reproduce just as you describe.

```
[oz@localhost code]$ ruby file.rb
/home/oz/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/minitest-5.10.1/lib/minitest/mock.rb:27:in `undef_method': undefined method `to_query' for class `Minitest::Mock' (NameError)
Did you mean?  to_query
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/minitest-5.10.1/lib/minitest/mock.rb:27:in `block in <class:Mock>'
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/minitest-5.10.1/lib/minitest/mock.rb:26:in `each'
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/minitest-5.10.1/lib/minitest/mock.rb:26:in `<class:Mock>'
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/minitest-5.10.1/lib/minitest/mock.rb:10:in `<module:Minitest>'
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/minitest-5.10.1/lib/minitest/mock.rb:3:in `<top (required)>'
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from /home/oz/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from file.rb:12:in `<main>''
```

**But** when I move `require 'minitest/mock` from the bottom of the script to above the module, it runs just fine. I don't see this error when I run Rails test suit, and can not find a file in Rails where 'minitest/mock' is required anywhere but at the top, together with the other requirements...

(Same result with ruby 2.4.0).

**Comment by zenspider** [comment]
OK. I finally got a repro. I'm also getting OTHER problems that aren't in the report:

```
7689 % ruby23 -vw -Ilib bug671.rb 
ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-darwin16]
/Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55: warning: loading in progress, circular require considered harmful - /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/deprecation/proxy_wrappers.rb
	from bug671.rb:2:in  `<main>'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/core_ext.rb:2:in  `<top (required)>'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/core_ext.rb:2:in  `each'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/core_ext.rb:3:in  `block in <top (required)>'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/core_ext/load_error.rb:1:in  `<top (required)>'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/deprecation/proxy_wrappers.rb:3:in  `<top (required)>'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/deprecation/proxy_wrappers.rb:4:in  `<module:ActiveSupport>'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/deprecation.rb:3:in  `<top (required)>'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/deprecation.rb:6:in  `<module:ActiveSupport>'
	from /Users/ryan/.gem/repos/minitest671/gems/activesupport-5.0.0/lib/active_support/deprecation.rb:19:in  `<class:Deprecation>'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in  `require'
/Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb:27:in `undef_method': undefined method `to_query' for class `Minitest::Mock' (NameError)
Did you mean?  to_query
	from /Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb:27:in `block in <class:Mock>'
	from /Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb:26:in `each'
	from /Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb:26:in `<class:Mock>'
	from /Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb:10:in `<module:Minitest>'
	from /Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb:3:in `<top (required)>'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Users/ryan/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from bug671.rb:12:in `<main>'
```

**Comment by zenspider** [comment]
@utilum wrt moving the require: makes sense, as whatever activerecord is doing to cause this problem hasn't been done yet. I'm working on figuring out what that is now.

**Comment by zenspider** [comment]
Minimal repro:

```ruby
class Object
  def to_query
  end
end

module WTF
end

Object.prepend WTF

module ObjectExtensions
  refine ::Object do
    def to_query
      'bar'
    end
  end
end

require 'minitest/mock'
```

**Comment by utilum** [comment]
👍 This is so minimal it looks like a bug in Ruby.

**Comment by zenspider** [comment]
I think so too:

https://bugs.ruby-lang.org/issues/13096

**Comment by erose** [comment]
Many thanks to you, @zenspider. I am amazed this turned out to be a bug in Ruby. Going ahead with a workaround for now.

**Comment by zenspider** [comment]
@erose can you put your workaround in here so others can find it?

**Comment by erose** [comment]
@zenspider I just removed the refinement. We were refining `Object` in the first place for not very good reasons :).

**Comment by zenspider** [comment]
This has been fixed in ruby and backported to 2.2 and 2.3!

https://bugs.ruby-lang.org/projects/ruby-trunk/repository/revisions/57362/diff

Closing. YAY
